lib/formatters/tapFormatter.mjs

```diff
@@ -57,3 +58,22 @@ export default function tapFormatter(results) {
 
 	return lines.join('\n');
 }
+
+/**
+ * replace octothorp/hash/hex/number sign/pound
+ * with its escape sequence
+ * in the file path or unique ID of the result's source
+ * @param {LintResult} result
+ * @modifies {result}
+ * @see {@link https://testanything.org/tap-version-14-specification.html#escaping TAP 14 specification}
+ */
+function escapeOctothorp(result) {
+	if (!result.source) return;
+
+	result.source = result.source.replaceAll('#', '%23');
```

Comment:
This seems incorrect.
Why is url encoding used and not `\` as dictated by the TAP spec?

Reply:
Because as explained in https://github.com/stylelint/stylelint/pull/8334#discussion_r1923831061:

1. it will ultimately be part of a path that may be copy/pasted or clicked on (not an id)
2. `%23` is `#`
3.  I consider that having a superfluous step is not necessary
i.e. `#` > `\#` > `%23`
4. we don't want to `encodeURIComponent` everything

Reply:
Do you have any supporting references that this is correct?
This is not according to the specification.

A TAP formatter is expected to be used with other tools that can interpret the TAP format.

Reply:
If you want we can add that step but I consider that it will require another step.
i.e. since it's a path we don't want to leave `\#` unencoded

Reply:
I think we are trying to solve different problems.

Currently we don't follow the TAP format.
Encoding as I described, as the specification describes is needed to correctly implement a TAP reporter.

uri encoding `#` makes is less correct, not more so.

What I think you are trying to achieve is that you can click on a TAP report text and open the corresponding file. Even when that TAP report contains a file name with `#` in the `description` field. I don't think that is correct given how TAP dictates that descriptions must be escaped.

Reply:
Let's take a real example to illustrate my point.

for `/home/lang/C#/foo/file.css` as the `source`
if we the formatter escapes `#` it now has the responsibility to output a valid URI
i.e. `/home/lang/C\#/foo/file.css` wouldn't be exploitable anymore
More over if it's an UNC file path which uses `\` instead of `/`.

What I mean is, I am not against escaping the characters, but if it introduces a regression it needs to be taken into account.
i.e. `\#` will have to be encoded back to `%23` if it's not a delimiter for the fragment identifier of the URI

So there's more to it than just a `replaceAll`, you were right about that.
Let's discuss this again once Ill make the changes that you have requested.